### PR TITLE
[#9536] Feedback Submit Page E2E Test

### LINKED ISSUE #9536: Add E2E testing to teammates pages in v7
Migrate end to end tests for all the pages below from v6 to v7

- [x] Student Home Page
- [x] Student Profile Page
- [x] Student Course Details Page
- [x] Admin Home Page
- [x] Admin Accounts Page
- [x] Admin Search Page
- [x] Admin Sessions Page
- [x] Instructor Home Page
- [x] Instructor Course Details Page
- [x] Instructor Courses Page
- [x] Instructor Course Edit Page
- [x] Instructor Course Enroll Page
- [x] Instructor Course Student Details Page
- [x] Instructor Course Student Edit Page
- [x] Instructor Search Page (ticked by default; see division of cases)
  - [x] Students table
  - [ ] Responses + response comments table
- [x] Instructor Student List Page
- [x] Instructor Student Records Page (ticked by default; see division of cases)
  - [x] Student details
  - [ ] Feedback responses given/received by student
- [x] Instructor Session Edit Page
- [x] Instructor Sessions Page
- [x] Instructor Session Result Page
- [x] Session Submission Page
- [x] Session Result Page
- [x] Course Join Page

### PR DESCRIPTION
**Pull request by jtankw3**
Part of #9536

Outline of Solution

- Add methods to BaseE2ETestCase to compare emails sent
- Add FeedbackSubmitPage to represent feedback submission page
- Add registration key to result of getStudentAction for backdoor requests.
- Add FeedbackSubmitPageE2ETest to test general actions on the submit page:
    - verify loaded session data
    - questions with giver type instructor
    - questions with giver type students
    - verify recipients: students
    - verify recipients: instructors
    - verify recipients: team mates
    - verify recipients: teams
    - submit partial response
    - cannot submit in closed session
    - can submit in grace period
    - confirmation email
    - add comment
    - edit comment
    - delete comment

Responding to each different specific question type will be tested in their own individual tests instead of here.

### CODE REVIEW

**Comment by wkurniawan07** [comment]
@jtankw3 don't forget to deconflict here

File: src/e2e/java/teammates/e2e/cases/e2e/FeedbackSubmitPageE2ETest.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+                + gracePeriodSession.getFeedbackSessionName() + "]");
+
+        ______TS("add comment");
+        int qnToComment = 1;
+        int recipientToComment = 1;
+        String comment = "<p>new comment</p>";
+        submitPage.addComment(qnToComment, recipientToComment, comment);
+
+        submitPage.verifyWarningMessageForPartialResponse(gracePeriodUnansweredQuestions);
+        // Reload page to verify comment persists in datastore
```
[the remark this answers is not in the document]
Comment (jtankw3) [reply src/e2e/java/teammates/e2e/cases/e2e/FeedbackSubmitPageE2ETest.java]: Added methods to use the existing `verifyPresentInDatastore`

**Comment by jtankw3** [comment]
As the previewing testcase in `InstructorFeedbackEditPageE2ETest` is failing intermittently due to some authentication problem when running multiple tests at once, I have simplified that test case and added all verification to `FeedbackSubmitPageE2ETest`

File: src/main/java/teammates/ui/webapi/action/GetStudentAction.java [hunk 2]
```diff
@@ -80,6 +80,9 @@ public ActionResult execute() {
         }
 
         StudentData studentData = new StudentData(student);
+        if (isBackdoor()) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/GetStudentAction.java:83]: I think instead of this, we can just check for admin privilege?

**Review by wkurniawan07 — COMMENTED** [review]
Mostly there, just one suggestion

File: src/main/java/teammates/ui/webapi/action/GetStudentAction.java [hunk 3]
```diff
@@ -80,6 +80,9 @@ public ActionResult execute() {
         }
 
         StudentData studentData = new StudentData(student);
+        if (userInfo.isAdmin) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/GetStudentAction.java:83]: Need to check that `userInfo` is not null

File: src/main/java/teammates/ui/webapi/action/Action.java [hunk 4]
```diff
@@ -95,7 +99,7 @@ public void checkAccessControl() {
     }
 
     private void initAuthInfo() {
-        if (Config.BACKDOOR_KEY.equals(req.getHeader("Backdoor-Key"))) {
+        if (isBackdoor()) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/Action.java:102]: This change is no longer necessary, I believe

**Review by wkurniawan07 — APPROVED** [review]
LGTM

**Pull request merged**